### ISSUE TITLE
setsockopt can wrongly detect OPTVAL parameter type when constants are used

### ISSUE THREAD

**Issue by Yaribz** [Perl/perl5#19892]
**Description**

As written in `setsockopt` documentation:

> setsockopt SOCKET,LEVEL,OPTNAME,OPTVAL
> OPTVAL might either be a packed string or an integer. An integer OPTVAL is shorthand for pack("i", OPTVAL).

So if a numeric constant is declared like this:

    use constant { RCVBUF_SIZE => 262144 };

And then `setsockopt` is called like this:

    setsockopt($sock,SOL_SOCKET,SO_RCVBUF,RCVBUF_SIZE);

Usually, it works fine....

However, if anywhere in some unrelated code the `RCVBUF_SIZE` constant is used to produce a string, even if this code is never actually executed, then the `setsockopt` call will fail to detect `RCVBUF_SIZE` as a number and won't perform the automatic `pack("i", OPTVAL)` call, which will lead to undesired result.

Of course it's still possible to use the `+0` trick to force number type, but this weird behavior can be very time-consuming to debug because the code using the constant to build a string can be very distant from the `setsockopt` call, and can even be dead code never actually executed.

**Steps to Reproduce**

Run this simple script once to check it works fine, then uncomment the last line, which just declares an unused function, and see how it prevents the script from working correctly.

```perl
use warnings;
use strict;

use Socket qw'PF_INET SOCK_STREAM IPPROTO_TCP SOL_SOCKET SO_RCVBUF';

use constant { RCVBUF_SIZE => 262144 };

socket(my $sock, PF_INET, SOCK_STREAM, IPPROTO_TCP)
  or die "Could not create socket - $!\n";

setsockopt($sock,SOL_SOCKET,SO_RCVBUF,RCVBUF_SIZE)
  or die "Could not set SO_RCVBUF on socket - $!\n";

my $rcvBuf=getsockopt($sock,SOL_SOCKET,SO_RCVBUF)
  or die "Could not get SO_RCVBUF on socket - $!\n";

$rcvBuf=unpack('i',$rcvBuf);

die "Unexpected SO_RCVBUF value: $rcvBuf\n"
  unless($rcvBuf == RCVBUF_SIZE || $rcvBuf == 2*RCVBUF_SIZE);

print "Ok.\n";
exit;

#sub bug {RCVBUF_SIZE.''}
```

**Perl configuration**

Tested on Linux with Perl 5.28 and Windows with Perl 5.32.

**Comment by xenu** [comment]
This is terrible but I'm not sure what we can do about it. Perhaps we should introduce a new feature flag that would make getsockopt() always take an integer.

**Comment by tonycoz** [comment]
Duplicates #18642.  The user is calling the op directly here, but they have the same underlying cause.

Given the changes in 5.36 we should be able to distinguish between the string and number forms now.  I'm only worried it may break working code.

**Comment by tonycoz** [comment]
fixed by the application of #18761
